Feature/2615 annotate pv with device name

slaperche-scality opened the pull request:
**Component**:

salt, operator

**Context**: 

In order to expose I/O metrics we need a way to link a PV/Volume (which is what we expose in our UI) to the underlying block device (which is the label used to export I/O metrics). 

**Summary**:

Add a new annotation on the PV created by our operator.
This annotation contains the name of the underlying block device and is refreshed at each reconciliation iteration (because it may change in some occasion, like a reboot).

**Acceptance criteria**: 

When you create a volume, the PV should have an annotation `storage.metalk8s.scality.com/device` providing the name of the underlying block device.
A test was added to check that and should pass.

---

Closes: #2615

slaperche-scality committed:
salt/volume: add a new module function to resolve device name

From a device path, it will return the name of the underlying block
device.

Refs: #2615
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
storage-operator/salt: use const to reduce duplication

Refs: #2615
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
storage-operator/salt: add submitJob helper

This reduce code duplication to submit a new Salt job, since most of the
code doesn't change between jobs (only the payload is variable).

Refs: #2615
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
storage-operator/salt: add GetDeviceName

Add a new job to resolve the device name of a volume.
This will be used to annotate a PV with the underlying device name.

Refs: #2615
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
storage-operator/controller: refresh device name on reconciliation

We use a synchronous call for this because:
- it makes the code simpler
- it's not critical to have this information (the volume is still
  working without it)
- we can easily retry it

Refs: 2615
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
tests/volumes: add a test to ensure PV is annotated with device name

Closes: #2615
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

bert-e commented:
# Hello slaperche-scality,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Branches have diverged



This pull request's source branch `feature/2615-annotate-pv-with-device-name` has diverged from
`development/2.6` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/2.6` into `feature/2615-annotate-pv-with-device-name`
* Rebase `feature/2615-annotate-pv-with-device-name` onto `origin/development/2.6`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/2615-annotate-pv-with-device-name` into
`development/2.5`
* [w/2.6/feature/2615-annotate-pv-with-device-name](https://github.com/scality/metalk8s/compare/development/2.6...w/2.6/feature/2615-annotate-pv-with-device-name) will be merged into `development/2.6` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @tcarmet

* @thomasdanan

* @sayf-eddine-scality

* @lucieleonard

* @monpote

* @carlito-scality

* @cmonfort

* @nicolast

* @slaperche-scality

* @teddyandrieux

* @nootal

* @gdemonet

* @alexandre-allard-scality

bert-e commented:
# History mismatch



Merge commit #38a05e95b9afc2f8711d1ae975b1cbf24604ec73 on the integration branch
`w/2.6/feature/2615-annotate-pv-with-device-name` is merging a branch which is neither the current
branch `feature/2615-annotate-pv-with-device-name` nor the development branch
`development/2.6`.

It is likely due to a rebase of the branch `feature/2615-annotate-pv-with-device-name` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

NicolasT commented on storage-operator/pkg/controller/volume/volume_controller.go:
```diff
@@ -201,6 +201,7 @@ reclaim its storage and remove the finalizers to let the object be deleted.
 
 }}} */
 
+const DEVICE_ANNOTATION = "storage.metalk8s.scality.com/device"
```
Considered `.../device-name`?

slaperche-scality replied:
Yup, I'll rename.

NicolasT commented on storage-operator/pkg/controller/volume/volume_controller.go:
```diff
@@ -713,6 +714,38 @@ func (self *ReconcileVolume) finalizeVolume(
+	volume *storagev1alpha1.Volume,
+	pv *corev1.PersistentVolume,
+) (reconcile.Result, error) {
+	path := pv.Spec.PersistentVolumeSource.Local.Path
```
Maybe have an assertion that the given PV is actually of type `Local`, i.e. `Path` is not `""`?

slaperche-scality replied:
It's supposed to be PV we created no?
But will do, it doesn't hurt to be extra-safe on that point.

NicolasT commented on storage-operator/pkg/controller/volume/volume_controller.go:
```diff
@@ -713,6 +714,38 @@ func (self *ReconcileVolume) finalizeVolume(
+		reqLogger.Info("update device name", "DeviceName", name)
+	}
+
+	return endReconciliation()
```
This implies this call is made at the end of a reconciliation loop, right? Which is not necessarily the case?

In general, shouldn't only `Reconcile` ever return a `reconcile.Result`, and (e.g.) this one an `error`, and let `Reconcile` translate this in a `delayedRequeue` or whatever?

slaperche-scality replied:
> This implies this call is made at the end of a reconciliation loop, right? Which is not necessarily the case?

Not necessarily.
It may be ill-named, but the role of this function is to return a `reconcile.Result` that end the reconciliation (i.e. doesn't trigger a requeue), unlike `requeue` and `delayedRequeue`.

> In general, shouldn't only `Reconcile` ever return a `reconcile.Result`, and (e.g.) this one an `error`, and let `Reconcile` translate this in a `delayedRequeue` or whatever?

Hum, that would be like a global try/catch, where we inspect the "exception" and decide how to react to it.
Could be done, but would need to define a bunch of error types to be able to tell the difference between the different sources of error.
Could be done, but deserve a dedicated issue.

NicolasT replied:
> > This implies this call is made at the end of a reconciliation loop, right? Which is not necessarily the case?
> 
> Not necessarily.
> It may be ill-named, but the role of this function is to return a `reconcile.Result` that end the reconciliation (i.e. doesn't trigger a requeue), unlike `requeue` and `delayedRequeue`.

I know. However, imagine this were *not* the last step/call in `Reconcile`. Now in order to know whether to continue with the next operation or ask for a re-enqueue without running the 'next step', `Reconcile` would need to look 'inside' this `reconcile.Result`...

> 
> > In general, shouldn't only `Reconcile` ever return a `reconcile.Result`, and (e.g.) this one an `error`, and let `Reconcile` translate this in a `delayedRequeue` or whatever?
> 
> Hum, that would be like a global try/catch, where we inspect the "exception" and decide how to react to it.
> Could be done, but would need to define a bunch of error types to be able to tell the difference between the different sources of error.
> Could be done, but deserve a dedicated issue.

Oh, for sure, out of scope.

Maybe the API could be as follows: instead of `(reconcile.Result, error)`, the functions could *still* return `(reconcile.Result, error)`, but the `Result` *must* be `nil` when `error` is `nil`, and may *not* be `nil` when `error` is not `nil`. As such, the `Result` is returned after the `error` is logged on error, otherwise `Reconcile` continues whatever it needs to do next, and in the end returns its own 'success' `Result`.

slaperche-scality replied:
Would be better yeah, I'll open a debt ticket for that.

slaperche-scality committed:
storage-operator: rename annotation to be more explicit

Use `device-name` instead of `device` to clarify that we return a name
and not a path.

Refs: #2615
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
storage-operator/controller: hand non-local PV

Shouldn't happen, but it's better to handle it (otherwise we could
crash).

Refs: #2615
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

NicolasT commented on storage-operator/pkg/controller/volume/volume_controller.go:
```diff
@@ -201,7 +201,7 @@ reclaim its storage and remove the finalizers to let the object be deleted.
 }}} */
 
-const DEVICE_ANNOTATION = "storage.metalk8s.scality.com/device"
+const DEVICE_ANNOTATION = "storage.metalk8s.scality.com/device-name"
```
Now thinking of it... Should this be an annotation, or rather in `status`? :thinking: WDYT @scality/metalk8s

gdemonet replied:
IMO, that's what `status` is for, and since we own it... I'd vote for using it. Feels minor however, and we could easily convert between one and the other in the future, if need be.

NicolasT replied:
Except that then the UI would need to dispatch between multiple possible locations :)

gdemonet replied:
Tie this schema to an `apiVersion`?

slaperche-scality replied:
Done, last commit move things into status.

gdemonet replied:
Oh thanks, was not sure if it needed to go in this PR.

slaperche-scality committed:
storage-operator: move device name into volume status

That way, everything is contained in the volume object.

Closes: #2615
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

gdemonet commented on storage-operator/pkg/controller/volume/volume_controller.go:
```diff
@@ -713,6 +713,40 @@ func (self *ReconcileVolume) finalizeVolume(
+	}
+	if volume.Status.DeviceName != name {
+		volume.Status.DeviceName = name
+		reqLogger.Info("update device name", "DeviceName", name)
```
I'm not familiar with how structured logging should work with controller-runtime, but given the `WithValues` call above, shouldn't we also namespace `DeviceName`?

```suggestion
		reqLogger.Info("update device name", "Volume.Status.DeviceName", name)
```

slaperche-scality replied:
I could "namespace" it yeah, I'll use `Volume.DeviceName` (I think the fact that this lives inside the Status sub-resource is an implementation detail, the one inspecting the log doesn't need this knowledge).

gdemonet commented on storage-operator/pkg/salt/client.go:
```diff
@@ -232,25 +205,73 @@ func (self *Client) GetVolumeSize(
+
+// Return the name of the block device designed by `devicePath`.
+//
+// This request is synchronous.
```
Back to using synchronous requests against SaltAPI? Isn't it a recipe for failure?

slaperche-scality replied:
As I tried to explain in the commit message, I think here it's acceptable because unlike `disk.dump` (which is in the "critical path" of volume creation):

- it doesn't prevent the volume to be used: a failed `disk.dump` can block/delay the creation of the volume, where a failed `metalk8s_volume.device_name` won't change the status/avaibility of the volume (volume is already available and will stay that way, worst case is we'll miss some state in the UI until the call succeed, but this doesn't prevent the volume to be ready and used).
- it's always retried on failure, no matter the "type" of error: unlike `disk.dump` (which only retry on some error and "abort" on some others), `metalk8s_volumes.device_name` is always retried on error

Given those difference, I think we can use a sync call (and have a big code simplification), without the drawbacks we used to have with `disk.dump`.

gdemonet replied:
Sorry I missed the commit message, OK for me!

gdemonet reviewed (dismissed):
Some minor questions, can safely be handled in follow-up tickets :)

bert-e commented:
# History mismatch



Merge commit #d28a5e4551dff10ea1accacf075759f5a8d06e07 on the integration branch
`w/2.6/feature/2615-annotate-pv-with-device-name` is merging a branch which is neither the current
branch `feature/2615-annotate-pv-with-device-name` nor the development branch
`development/2.6`.

It is likely due to a rebase of the branch `feature/2615-annotate-pv-with-device-name` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

slaperche-scality commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.6/feature/2615-annotate-pv-with-device-name` with contents from `feature/2615-annotate-pv-with-device-name`
and `development/2.6`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.6/feature/2615-annotate-pv-with-device-name origin/development/2.6
 $ git merge origin/feature/2615-annotate-pv-with-device-name
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.6/feature/2615-annotate-pv-with-device-name
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @tcarmet

* @thomasdanan

* @sayf-eddine-scality

* @lucieleonard

* @monpote

* @carlito-scality

* @cmonfort

* @nicolast

* @slaperche-scality

* @teddyandrieux

* @nootal

* @gdemonet

* @alexandre-allard-scality

slaperche-scality commented:
/approve

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/be8f6ca15c1e56a5d1a4ba96b0894ed28e84e1e5) did not succeed in branch w/2.6/feature/2615-annotate-pv-with-device-name.




*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/e90f40f21f6668202bda61e04f121acf0be610c0) did not succeed in branch feature/2615-annotate-pv-with-device-name.




*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/be8f6ca15c1e56a5d1a4ba96b0894ed28e84e1e5) did not succeed in branch w/2.6/feature/2615-annotate-pv-with-device-name.




*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/e90f40f21f6668202bda61e04f121acf0be610c0) did not succeed in branch feature/2615-annotate-pv-with-device-name.




*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/be8f6ca15c1e56a5d1a4ba96b0894ed28e84e1e5) did not succeed in branch w/2.6/feature/2615-annotate-pv-with-device-name.




*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/e90f40f21f6668202bda61e04f121acf0be610c0) did not succeed in branch feature/2615-annotate-pv-with-device-name.




*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/be8f6ca15c1e56a5d1a4ba96b0894ed28e84e1e5) did not succeed in branch w/2.6/feature/2615-annotate-pv-with-device-name.




*The following options are set:* **approve**

slaperche-scality committed:
salt: add missing permissions for the storage operator

Otherwise we can't call the new module function.

Refs: #2615
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @tcarmet

* @thomasdanan

* @sayf-eddine-scality

* @lucieleonard

* @monpote

* @carlito-scality

* @cmonfort

* @nicolast

* @slaperche-scality

* @teddyandrieux

* @nootal

* @gdemonet

* @alexandre-allard-scality










*The following options are set:* **approve**

slaperche-scality commented on salt/metalk8s/salt/master/files/master-99-metalk8s.conf.j2:
```diff
@@ -45,6 +45,7 @@ external_auth:
     storage-operator:
       - '*':
         - 'disk.dump'
+        - 'metalk8s_volumes.device_name'
```
Weirdly enough, it was working fine in 2.5 without that, but was failing in 2.6 with a 401 error from SaltAPI (as expected, since the permission was missing).

gdemonet replied:
That's a bug, here's a ticket to investigate: #2634

bert-e commented:
# History mismatch



Merge commit #867c0b9baaa5ba4c06c3c3a196139da3993c524d on the integration branch
`w/2.6/feature/2615-annotate-pv-with-device-name` is merging a branch which is neither the current
branch `feature/2615-annotate-pv-with-device-name` nor the development branch
`development/2.6`.

It is likely due to a rebase of the branch `feature/2615-annotate-pv-with-device-name` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

slaperche-scality commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.6/feature/2615-annotate-pv-with-device-name` with contents from `feature/2615-annotate-pv-with-device-name`
and `development/2.6`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.6/feature/2615-annotate-pv-with-device-name origin/development/2.6
 $ git merge origin/feature/2615-annotate-pv-with-device-name
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.6/feature/2615-annotate-pv-with-device-name
```





*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @tcarmet

* @thomasdanan

* @sayf-eddine-scality

* @lucieleonard

* @monpote

* @carlito-scality

* @cmonfort

* @nicolast

* @slaperche-scality

* @teddyandrieux

* @nootal

* @gdemonet

* @alexandre-allard-scality










*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.5`

* :heavy_check_mark: `development/2.6`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

slaperche-scality commented:
/approve

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.5`

* :heavy_check_mark: `development/2.6`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`



Please check the status of the associated issue None.

Goodbye slaperche-scality.
